src/coreclr/jit/fgehopt.cpp

```diff
@@ -659,7 +759,298 @@ PhaseStatus Compiler::fgRemoveEmptyTry()
+            BasicBlock* const afterLastFltBlock = HBtab->BBFilterLast()->Next();
+
+            // Must do this in two passes to handle loops or lexically
+            // backwards references.
+            //
+            for (BasicBlock* filterBlock = firstFltBlock; filterBlock != afterLastFltBlock;
+                 filterBlock             = filterBlock->Next())
+            {
+                fgRemoveBlockAsPred(filterBlock);
+                filterBlock->SetKind(BBJ_THROW);
+            }
+
+            for (BasicBlock* filterBlock = firstFltBlock; filterBlock != afterLastFltBlock;
+                 filterBlock             = filterBlock->Next())
+            {
+                filterBlock->RemoveFlags(BBF_DONT_REMOVE);
+                fgRemoveBlock(filterBlock, /* unreachable */ true);
+            }
+        }
+
+        // (3) Remove any handler blocks.
+        // The first handler block has an artificial ref count
+        //
+        assert(firstHndBlock->bbRefs == 1);
+        firstHndBlock->bbRefs               = 0;
+        BasicBlock* const afterLastHndBlock = lastHndBlock->Next();
+
+        // Must do this in two passes to handle loops or lexically
+        // backwards references.
+        //
+        for (BasicBlock* handlerBlock = firstHndBlock; handlerBlock != afterLastHndBlock;
+             handlerBlock             = handlerBlock->Next())
+        {
+            assert(!bbIsTryBeg(handlerBlock));
+            fgRemoveBlockAsPred(handlerBlock);
+            handlerBlock->SetKind(BBJ_THROW);
+        }
+
+        for (BasicBlock* handlerBlock = firstHndBlock; handlerBlock != afterLastHndBlock;
+             handlerBlock             = handlerBlock->Next())
+        {
+            assert(!bbIsTryBeg(handlerBlock));
+            handlerBlock->RemoveFlags(BBF_DONT_REMOVE);
+            fgRemoveBlock(handlerBlock, /* unreachable */ true);
+        }
+
+        // (4) Update any impacted ACDs.
+        //
+        fgUpdateACDsBeforeEHTableEntryRemoval(XTnum);
+
+        // (5) Remove the try-catch EH region. This will compact the
+        // EH table so XTnum now points at the next entry and will update
+        // the EH region indices of any nested EH blocks.
+        //
+        fgRemoveEHTableEntry(XTnum);
+
+        // (6) The old try entry no longer needs special protection.
+        //
+        firstTryBlock->RemoveFlags(BBF_DONT_REMOVE);
+
+        // Another one bites the dust...
+        emptyCount++;
+    }
+
+    if (emptyCount > 0)
+    {
+        JITDUMP("fgRemoveEmptyTryCatch() optimized %u empty-try catch clauses\n", emptyCount);
+
+        const bool hadDfs   = (m_dfsTree != nullptr);
+        const bool hadLoops = (m_loops != nullptr);
+
+        if (hadDfs)
+        {
+            fgInvalidateDfsTree();
+            m_dfsTree = fgComputeDfs();
+            if (hadLoops)
+            {
+                m_loops = FlowGraphNaturalLoops::Find(m_dfsTree);
+            }
+        }
```

Comment:
It seems unnecessary to recompute this in each of the phases. Maybe combine it into one phase, or make the actual phase that needs these annotations compute them if they aren't computed when we get there?

Reply:
moved recomputation later (to doms for dfs, vn for loops).